### ISSUE TITLE
std::filesystem::path is not supported in macOS<10.15

### ISSUE THREAD

**Issue by sdatkinson** [sdatkinson/NeuralAmpModelerPlugin#23]
[Use of `std::filesystem::path`](https://github.com/sdatkinson/NeuralAmpModelerPlugin/blob/31e69c8e86d41f09ca96317e89e87d3e2ad40185/NeuralAmpModeler/NeuralAmpModeler.cpp#L292) is preventing support for macOS versions before 10.15.

**Comment by masqutti** [comment]
Did some research on this and based on https://en.cppreference.com/w/cpp/filesystem std::filesystem originally was a feature of boost library so https://www.boost.org/doc/libs/1_81_0/libs/filesystem/doc/index.htm ... Maybe include that in the project instead for backwards compatibility? Or then go with some native route. :)

**Comment by sdatkinson** [comment]
Might be able to solve this using `WDL_String`s from iPlug2.

**Comment by sdatkinson** [comment]
Drawback would be that the dsp module would be locked into iPlug2, which would be bad.

**Comment by Moonheadmix** [comment]
Looking forward to a fix to be able to use in Mojave thank you

**Comment by masqutti** [comment]
Just found this which might be helpful! https://github.com/gulrak/filesystem

"An implementation of C++17 std::filesystem for C++11 /C++14/C++17/C++20 on Windows, macOS, Linux and FreeBSD."

**Comment by Gordonozzy** [comment]
Mojave 10.14 user—Curious if this is still in progress? Thanks

**Comment by masqutti** [comment]
> Just found this which might be helpful! https://github.com/gulrak/filesystem
> 
> "An implementation of C++17 std::filesystem for C++11 /C++14/C++17/C++20 on Windows, macOS, Linux and FreeBSD."

According to my tests gulrak's library is a single-header with handy macro helpers and I managed to build nam plugin with it and C++14. Unfortunately it seems that iPlug2 still requires C++17 but there was just a couple of lines to fix so that built with C++14 too.

**Comment by masqutti** [comment]
https://github.com/masqutti/NeuralAmpModelerPlugin/commit/e3fa057f1f110b8d92dabd70d00f04a3415eed1c

Here's info how I did it, it's not a big change but I unfortunately too occupied at the moment to do pull requests myself. It requires the https://github.com/gulrak/filesystem to exist and subfolder "Include" added as to header search paths to allow building. Also iPlug2 needs two small changes in couple of places (you will notice when you try to build) which are about Structured bindings (not possible with C++14) and very easy to fix, probably ideal to have iPlug2 issue / pull request about them since compiler only complained about 2 lines.

**Comment by mikeoliphant** [comment]
I think it would be better to just pass paths as strings, rather than add another dependency with https://github.com/gulrak/filesystem.

From looking at the code, it isn't doing much specific to std::filesystem::path beyond some file existence checks.

**Comment by masqutti** [comment]
Oh yes definitely, if strings are enough. I guess I'll investigate it a bit further when I got the time 🤔

**Comment by EchoReverbFilter** [comment]
> 期待能够在 Mojave 中使用的修复程序，谢谢
me 2, really want try NAM on MAC 10.14 Mojave ! if that works someday, please tell me . thank u !

**Comment by EchoReverbFilter** [comment]
> me 2, really want try NAM on MAC 10.14 Mojave ! if that works someday, please tell me . thank u !

sorry i dont know anything about coding.... im not a geek but really apreaciate ur work , hope u cant work it out ! really want try NAM on MAC 10.14 Mojave ! if that works someday, please tell me . thank u !

**Comment by Vladzibor** [comment]
Oh, what a pity that the plugin does not work in MOJAVE 10.13.6 Millions of users still work in this operating system. Especially in such difficult times.

**Comment by tnyeanderson** [comment]
Hey @masqutti, any chance this could be revisited? Now that the iPlug2 fix is merged, perhaps this has become more straightforward.

I tried to make a PR including the changes from your commit linked above, but there's a lot of merge conflicts and it's a bit hard to follow the history and separate what is still needed since I am unfamiliar with the codebase.

I'm willing to help where I can, and I'm happy to test on my 10.13 Mac :)

**Comment by masqutti** [comment]
> Hey @masqutti, any chance this could be revisited? Now that the iPlug2 fix is merged, perhaps this has become more straightforward.
> 
> I tried to make a PR including the changes from your commit linked above, but there's a lot of merge conflicts and it's a bit hard to follow the history and separate what is still needed since I am unfamiliar with the codebase.
> 
> I'm willing to help where I can, and I'm happy to test on my 10.13 Mac :)

Yes iPlug removed the tiny C++17 feature and it should work, I haven't checked in for a while tho. But NAM still includes those C++17 filesystem and clamp features that prevent using NAM on older Mac OS version. clamp is an easy one, it's a simple limiting of a value inside a range, but std::filesystem features at least exists() method which checks if a file exists or not, which is required for NAM to work and thus not simple std::strings can work alone to fix this. I have forks which solved these issues and I've built NAM for down to Mac OS 10.9, but they're not up to date atm:

https://github.com/masqutti/NeuralAmpModelerPlugin/tree/dspStruct
https://github.com/masqutti/NeuralAmpModelerCore/tree/dspStruct

It seems gulrak's filesystem which I've used to replace std::filesystem has been updated 2 weeks ago so it's still an active repo, to me it's still worth considering.

**Comment by masqutti** [comment]
@sdatkinson I can definitely still do a pull request which includes gulrak's filesystem, either as a submodule or just plain ol copy/paste ;) But I won't attempt if it's not your intention to include more repos into NAM.

**Comment by tnyeanderson** [comment]
Based on how many submodules are already in this repo, I think it's fair to assume that it's an acceptable method :)

I think, if it's not too inconvenient for you, getting a PR opened with an up-to-date solution to this problem will at least allow a review and give us a starting point. If changes need to be made before the PR is accepted, anyone can help out since we can all easily check the diff to see exactly what is needed to remove this dependency and allow the software to work on older systems.

Thanks for the quick response and for your hard work on this project!

**Comment by sdatkinson** [comment]
Thanks all for your patience on this.

I'm open to either of the two paths forward:
* using gulrak's filesystem
* Implementing the needed methods first-party using `std:string`s.

I think slightly prefer gulrak on account of (1) it having tests, (2) this feels like a potentially-better solution for the long run if other file manipulation comes into view that would benefit from already having a more fleshed-out implementation, and (3) the license is fine and dandy, and (4) besides, if this isn't much work, I'm sure it can be rethought in the future if need be (and keeping the `std` interfaces sounds nice).

So if anyone wants to PR me this I'll give it to whomever raises their hand first :)

**Comment by tommyq94** [comment]
Hey everyone,

Is NAM able to load in Mojave, or will this ever be possible? 

I am clueless as to coding etc, but I really wanted to try out the NAM plug-in.

Would really appreciate any help with this!

Thanks

**Comment by masqutti** [comment]
> Thanks all for your patience on this.
> 
> I'm open to either of the two paths forward:
> 
> * using gulrak's filesystem
> * Implementing the needed methods first-party using `std:string`s.
> 
> I think slightly prefer gulrak on account of (1) it having tests, (2) this feels like a potentially-better solution for the long run if other file manipulation comes into view that would benefit from already having a more fleshed-out implementation, and (3) the license is fine and dandy, and (4) besides, if this isn't much work, I'm sure it can be rethought in the future if need be (and keeping the `std` interfaces sounds nice).
> 
> So if anyone wants to PR me this I'll give it to whomever raises their hand first :)

I could work on this one a bit, as I already have implemented gulrak's on my repository. I wonder where should I put the submodule, in NeuralAmpModelerCore's dependencies -folder or in this repository? I noticed there's eigen in both. :o

**Comment by dw199** [comment]
As a producer still working on an older system, is there a download/fork that will allow NAM to run on 10.13? Have been trying to figure out a way to use NAM.

Appreciate all the hard work that has gone into this :-)

**Comment by sdatkinson** [comment]
@masqutti I think I'd prefer putting it here and not NeuralAmpModelerCore. It'd be easier to move it to core later than the opposite I think.

**Comment by sdatkinson** [comment]
@masqutti I've assigned you to this since it sounds like you shouldn't have a problem getting it done; lmk if that's ok

**Comment by masqutti** [comment]
> @masqutti I've assigned you to this since it sounds like you shouldn't have a problem getting it done; lmk if that's ok

It's ok! :)

**Comment by josmithiii** [comment]
Hi All - I hit this as well and here are my diffs to address it:

```
<.../NeuralAmpModeler/NeuralAmpModelerCore> 
diff --git a/NAM/dsp.h b/NAM/dsp.h
index 2f88cbd..ad31999 100644
--- a/NAM/dsp.h
+++ b/NAM/dsp.h
@@ -6,6 +6,7 @@
 #include <string>
 #include <unordered_map>
 #include <vector>
+#include <filesystem>

 #include <Eigen/Dense>
```

**Comment by dw199** [comment]
Hi all, Any updates about potential 10.13 compatibility? Would love to use this awesome piece of software on a project I'm starting!!

**Comment by masqutti** [comment]
> Hi all, Any updates about potential 10.13 compatibility? Would love to use this awesome piece of software on a project I'm starting!!

@sdatkinson I'm sorry for being laz.. late! Well I looked it today and just a _heads up_: As this filesystem modification is about C++14 compatibility, there's a couple of other things to change too. std::clamp and std::string_view are part of C++17. They're trivial to change but this means that there's a pull request incoming for 3 repositories... this one, NeuralAmpModelerCore and AudioDSPTools, which all need tiny adjustments to allow C++14 compatibility.
